benchmark/benchmark_utils.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Benchmarking utilities.
+"""
+import abc
+
+import pennylane as qml
+
+
+
+class BaseBenchmark(abc.ABC):
+    """ABC for benchmarks.
+
+    Args:
+        device (~pennylane.Device): device for executing the benchmark (if needed)
+        verbose (bool): If True, print debugging info during the benchmark. Note that this
+            may invalidate the benchmark due to printing latency that should be irrelevant.
+    """
+    name = None  #: str: benchmark name
+    min_wires = 1  #: int: minimum number of quantum wires required by the benchmark
+    n_vals = None  #: Sequence[Any]: range of benchmark parameter values for perfplot
+
+    def __init__(self, device=None, verbose=False):
+        self.device = device
+        if device is not None:
+            if device.num_wires < self.min_wires:
+                raise ValueError("'{}' requires at least {} wires.".format(self.name, self.min_wires))
+            self.n_wires = device.num_wires
+        else:
+            self.n_wires = None
+        self.verbose = verbose
+
+    @abc.abstractmethod
+    def setup(self):
+        """Set up the benchmark.
+
+        This method contains the initial part of the benchmark that should not be timed.
+        """
+
+    @abc.abstractmethod
+    def benchmark(self, n):
+        """The benchmark test itself.
+
+        Args:
+            n (int): benchmark size parameter
+
+        Returns:
+            Any: Result of the benchmark. Not None.
```

Comment:
Does `Any` not usually include `None`? Why not `None`?

Reply:
Yeah, Any can mean None too. I think perfplot requires that the kernels return something (and that they all return the same value, for a correctness check).

Reply:
```suggestion
            Any: Result of the benchmark. Must not return None.
```